File: go/src/app/lotherry.go
package main

import (
    "github.com/fedesog/webdriver"
	"log"
	"time"
	"math/rand"
	"os"
	"encoding/csv"
	"strconv"
)


//CHROME_DRIVER_LOCATION
const DefaultDriverLocation = "chromedriver"
//SCRAPPING_URL
const DefaultScrappingUrl = ""
//OUTPUT_FILE
const DefaultOutputFile = "winning_numbers"
//MAX_RETRIES
const DefaultRetryAttempts = "20"

func main() {

	chromeDriverLocation := getEnvironmentVariable("CHROME_DRIVER_LOCATION", DefaultDriverLocation)
	chromeDriver := webdriver.NewChromeDriver(chromeDriverLocation)
	err := chromeDriver.Start()
	checkError("Cannot start driver", err)

	chromeOptions := make(map[string][]string)
	chromeOptions["args"] = []string{"--headless", "--disable-gpu", "--window-size=1280,800"}
	desired := webdriver.Capabilities{"Platform": "Linux", "chromeOptions" : chromeOptions}

	required := webdriver.Capabilities{}
	session, err := chromeDriver.NewSession(desired, required)
	checkError("Cannot create session", err)
	time.Sleep(10 * time.Second)
	url := getEnvironmentVariable("SCRAPPING_URL", DefaultScrappingUrl)
	err = session.Url(url)
	checkError("Cannot fetch url", err)

	timestamp := time.Now().Format("2006-01-02T15:04:05")
	outputFile := getEnvironmentVariable("OUTPUT_FILE", DefaultOutputFile) + "_" + timestamp
	file, err := os.Create(outputFile)
	checkError("Cannot create file", err)
	defer file.Close()

	writer := csv.NewWriter(file)
	defer writer.Flush()

	csvRow := []string{"privateKey", "address", "balance"}
	err = writer.Write(csvRow)
	checkError("Cannot write to file", err)

	maxRetries, err := strconv.Atoi(getEnvironmentVariable("MAX_RETRIES", DefaultRetryAttempts))
	checkError("Cannot parse max retries", err)

	for 1 == 1 {
		log.Println("Fetching next page")
		err = session.Url(url)
		//checkError("Cannot fetch url", err)
		session.AcceptAlert()

		rows, err := session.FindElements(webdriver.FindElementStrategy("tag name"), "tr")
		checkError("Cannot fetch records", err)

		for rowIndex := 1; rowIndex < len(rows); rowIndex++ {
			rowData, err := rows[rowIndex].FindElements(webdriver.FindElementStrategy("tag name"), "td")
			checkError("Cannot fetch row data", err)

			for attempt := 1; attempt < maxRetries; attempt++ {
				rowInfo := parseRow(rowData)
				balance := rowInfo["balance"]

				if len(balance) == 0 {
					log.Printf("Attempt %d of %d. Balance not ready. Retrying", attempt, maxRetries)
					continue
				}
				if len(balance) != 0 {
					log.Printf("%s %s %s", rowInfo["privateKey"], rowInfo["address"], balance)
					if balance != "0" {
						log.Printf("Bingo: %s %s %s %s", rowInfo["privateKey"], rowInfo["address"], rowInfo["balance"], balance)
						csvRow := []string{rowInfo["privateKey"], rowInfo["address"], balance}
						err = writer.Write(csvRow)
						checkError("Cannot write to file", err)
						writer.Flush()
					}
					break
				}
				if attempt == maxRetries - 1 {
					log.Printf("Exhausted all attempts for row %d", rowIndex)
				}
			}
		}
		var seconds = random(5, 10)
		log.Printf("Sleeping for %d", seconds)
		time.Sleep(time.Duration(seconds) * time.Second)
	}

	session.Delete()
	chromeDriver.Stop()
}

func getEnvironmentVariable(key string, defaultValue string) (string) {
	variableValue := os.Getenv(key)
	if len(variableValue) == 0 {
		log.Printf("No env variable specified for %s. Using default: %s", key, defaultValue)
		return defaultValue
	} else {
		log.Printf("Using env variable %s=%s", key, variableValue)
		return variableValue
	}
}

func parseRow(rowData []webdriver.WebElement)(map[string]string){
	var rowInfo = make(map[string]string)
	for column := range rowData {
		text, textFetchErr := rowData[column].Text()
		if textFetchErr != nil {
			log.Println(textFetchErr)
		}
		if column == 2 {
			rowInfo["balance"] = text
		} else if column == 0 {
			rowInfo["privateKey"] = text
		} else if column == 1 {
			rowInfo["address"] = text
		}
	}
	return rowInfo
}

func checkError(message string, err error) {
	if err != nil {
		log.Fatal(message, err)
	}
}

func random(min, max int) int {
	rand.Seed(time.Now().Unix())
	return rand.Intn(max - min) + min
}

File: lotherry.py
import logging
import os
import time

from selenium import webdriver
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.by import By
from selenium.common.exceptions import NoAlertPresentException

logging.basicConfig(level=logging.INFO,
        format='%(asctime)s [%(levelname)s] (%(threadName)-10s) %(message)s',
        )


MAX_ATTEMPTS = int(os.getenv('MAX_ATTEMPTS', 50))
OUTPUT_FILE = os.getenv('OUTPUT_FILE', 'winning_numbers')

def acceptAlert(driver):
    try:
        time.sleep(0.02)
        alert = driver.switch_to.alert
        alert.accept()
        logging.info("Alert accepted")
    except NoAlertPresentException:
        a = 1

def main():
    logging.info("Configuring options...")
    options = webdriver.ChromeOptions()
    options.add_argument('headless')
    options.add_argument('window-size=1200x600')
    logging.info("Starting webdriver...")
    driver = webdriver.Chrome(chrome_options=options)

    output = open('file_{}_{}'.format(OUTPUT_FILE, int(time.time())), 'w')
    output.write("      key     |       address     |       balance     \n")
    output.flush()

    while True:
        logging.info("Fetching page...")
        driver.get("http://www.ethersecret.com/random")
        acceptAlert(driver)
        rows = driver.find_elements(By.TAG_NAME, "tr")
        acceptAlert(driver)
        for rowIdx, row in enumerate(rows):
            if (rowIdx == 0):
                continue
            columns = row.find_elements(By.TAG_NAME, "td")
            acceptAlert(driver)
            for attempt in range(1, MAX_ATTEMPTS):
                logging.info("Attempt %d of %d", attempt, MAX_ATTEMPTS)
                rowInfo = {}
                for idx, val in enumerate(columns):
                    acceptAlert(driver)
                    if (idx == 0):
                        rowInfo['key'] = val.text
                        acceptAlert(driver)
                    if (idx == 1):
                        rowInfo['address'] = val.text
                        acceptAlert(driver)
                    if (idx == 2):
                        rowInfo['balance'] = val.text
                        acceptAlert(driver)
                try:
                    balance = rowInfo['balance']
                    if (balance == "" or balance == None):
                        continue
                    logging.info("%s | %s | %s", rowInfo['key'], rowInfo['address'], rowInfo['balance'])
                    if (balance != "0"):
                        logging.info("Bingo: %s | %s | %s", rowInfo['key'], rowInfo['address'], rowInfo['balance'])
                        output.write(rowInfo['key'] + " " + rowInfo['address']
                                + " " + rowInfo['balance'] + "\n")
                        output.flush()
                    break
                except KeyError:
                    time.sleep(1)
                    continue
                logging.info("%s | %s | %s", rowInfo['key'], rowInfo['address'], rowInfo['balance'])

if __name__ == "__main__":
    main()
